crates/matrix-sdk/src/room/timeline/mod.rs

```diff
@@ -164,29 +166,66 @@ impl Timeline {
     }
 
     /// Add more events to the start of the timeline.
-    #[instrument(skip(self), fields(room_id = %self.room.room_id()))]
-    pub async fn paginate_backwards(&self, limit: UInt) -> Result<PaginationOutcome> {
+    ///
+    /// # Arguments
+    ///
+    /// * `initial_pagination_size`: The number of events to fetch from the
+    ///   server in the first pagination request. The server may choose return
+    ///   fewer events, for example because the supplied number is too big or
+    ///   the beginning of the visible timeline was reached.
+    /// * `
+    #[instrument(skip_all, fields(initial_pagination_size, room_id = %self.room.room_id()))]
+    pub async fn paginate_backwards(&self, mut opts: PaginationOptions<'_>) -> Result<()> {
         let mut start_lock = self.start_token.lock().await;
+        if start_lock.is_none()
+            && self.inner.items.lock_ref().first().map_or(false, |item| item.is_timeline_start())
+        {
+            warn!("Start of timeline reached, ignoring backwards-pagination request");
+            return Ok(());
+        }
+
         self.inner.add_loading_indicator();
-        let messages = self
-            .room
-            .messages(assign!(MessagesOptions::backward(), {
-                from: start_lock.clone(),
-                limit,
-            }))
-            .await?;
 
         let own_user_id = self.room.own_user_id();
-        let mut num_updates = 0;
-        for room_ev in messages.chunk {
-            num_updates += self.inner.handle_back_paginated_event(room_ev, own_user_id).await;
+        let mut from = start_lock.clone();
+        let mut pagination_outcome = PaginationOutcome::new();
+
+        while let Some(limit) = opts.next_event_limit(pagination_outcome) {
+            let messages = self
+                .room
+                .messages(assign!(MessagesOptions::backward(), {
+                    from,
+                    limit: limit.into(),
+                }))
+                .await?;
+
+            pagination_outcome.events_received = messages
+                .chunk
+                .len()
+                .try_into()
+                .expect("Pagination response can't have so >65535 events");
```

Comment:
Can't we use an usize for `events_received` to avoid this expect? 

I understand that this is unlikely to be a problem but at the same time it isn't obvious to me that `chunk.len()` has to fit into a `u16` or that a buggy/malicious server can't trigger a crash here.